Atomic/syscontainers.py

```diff
@@ -386,6 +386,147 @@ def _write_config_to_dest(self, destination, exports_dir, values=None):
         else:
             self._generate_default_oci_configuration(destination)
 
+    @staticmethod
+    def _get_manifest_attributes(manifest, attr_name, default_value):
+        """
+        Collects corresponding information from manifest file
+        based on attribtue name
+
+        :returns attribtues' corresponding value if it does exist
+        else returns a default value that is predefined
+        """
+        if manifest is not None and attr_name in manifest:
+            return manifest[attr_name]
+        return default_value
+
+    def _upgrade_tempfiles_clean(self, manifest, options, was_service_active):
+        """
+        When upgrading a container, we stop the service and remove previously
+        installed tmpfiles, before restarting the service
+
+        :param manifest: dictionary loaded from manifest json file
+        :param options: a dictionary which contains user input collectively
+        :param was_service_active: a boolean to indicate whether the container service was active
+        """
+        # The thought process here might be a bit tricky. When noContainerService is true, has_container_service
+        # will have the opposite value. Thus goes with the 'not' sign. Then because of the 'not', the default return
+        # value when noContainerService is not present should be False so that has_container_service will have
+        # a default value of True
+        has_container_service = not(SystemContainers._get_manifest_attributes(manifest, "noContainerService", False))
+
+        if has_container_service and options["upgrade_mode"] != SystemContainers.CHECKOUT_MODE_INSTALL:
+            if was_service_active:
+                self._systemctl_command("stop", options["name"])
+            if os.path.exists(options["tmpfilesout"]):
+                try:
+                    self._systemd_tmpfiles("--remove", options["tmpfilesout"])
+                except subprocess.CalledProcessError:
+                    pass
+
+    def _update_rename_file_value(self, manifest, values):
+        """
+        Substitute entries from rename_files content in manifest by the generated input values
+
+        :param manifest: dictionary loaded from manifest json file
+        :param values: the generated input values
+        """
+        rename_files = SystemContainers._get_manifest_attributes(manifest, "renameFiles", {})
+
+        if rename_files:
+            for k, v in rename_files.items():
+                template = Template(v)
+                try:
+                    new_v = template.substitute(values)
+                except KeyError as e:
+                    raise ValueError("The template file 'manifest.json' still contains an unreplaced value for: '%s'" % \
+                                     (str(e)))
+                rename_files[k] = new_v
+
+    def _handle_system_package_files(self, options, manifest, exports):
```

Comment:
Need help/hints on writing unit test for RPMHostInstall related functions.. I kinda "give up" after trying a while..

Reply:
If you are having issues because of side effects look at ``mock`` as a way of intercepting and faking responses back to code.

Reply:
Note: I specifically mean Python mock and not Fedora mock (different Python lib/tool)